manifests/init.pp

```diff
@@ -1,54 +1,75 @@
 # == Class: puppet_agent
 #
-# Upgrades Puppet 3.8 to Puppet 4+ (Puppet-Agent from Puppet Collection 1).
-# Makes the upgrade easier by migrating SSL certs and config files to the new
-# Puppet-Agent paths and removing deprecated settings that are no longer
+# Upgrades Puppet 3.8 and newer to the requested version.
+# Makes Puppet 4 upgrades easier by migrating SSL certs and config files to the
+# new Puppet-Agent paths and removing deprecated settings that are no longer
 # supported by Puppet 4.
 #
 # === Parameters
 #
+# [arch]
+#   The package architecture. Defaults to the architecture fact.
+# [collection]
+#   The Puppet Collection to track. Defaults to 'PC1'.
+# [is_pe]
+#   Install from Puppet Enterprise repos. Enabled if communicating with a PE master.
 # [package_name]
 #   The package to upgrade to, i.e. `puppet-agent`.
+# [package_version]
+#   The package version to upgrade to. Defaults to undef, meaning only upgrade from Puppet < 4.0.
 # [service_names]
 #   An array of services to start, normally `puppet` and `mcollective`.
 #   None will be started if the array is empty.
+# [source]
+#   The location to find packages.
 #
 class puppet_agent (
-  $arch          = $::architecture,
-  $is_pe         = $::puppet_agent::params::_is_pe,
-  $package_name  = $::puppet_agent::params::package_name,
-  $service_names = $::puppet_agent::params::service_names,
-  $source        = $::puppet_agent::params::_source,
+  $arch            = $::architecture,
+  $collection      = 'PC1',
+  $is_pe           = $::puppet_agent::params::_is_pe,
+  $package_name    = $::puppet_agent::params::package_name,
+  $package_version = undef,
+  $service_names   = $::puppet_agent::params::service_names,
+  $source          = $::puppet_agent::params::_source,
 ) inherits ::puppet_agent::params {
 
   validate_re($arch, ['^x86$','^x64$','^i386$','^i86pc$','^amd64$','^x86_64$','^power$','^sun4[uv]$','PowerPC_POWER'])
 
-  if versioncmp("${::clientversion}", '3.8.0') < 0 {
-    fail('upgrading requires Puppet 3.8')
+  if ($package_version == undef and versioncmp("${::clientversion}", '4.0.0') < 0) {
+    $_package_version = $puppet_agent::params::master_agent_version
+  } elsif ($package_version != undef and versioncmp("${::clientversion}", $package_version) != 0) {
+    $_package_version = $package_version
   }
-  elsif versioncmp("${::clientversion}", '4.0.0') >= 0 {
-    info('puppet_agent performs no actions on Puppet 4+')
+
+  if versioncmp("${::clientversion}", '3.8.0') < 0 {
+    fail('upgrading requires at least Puppet 3.8')
   }
-  else {
+  elsif $is_pe and $::aio_agent_version != undef and $_package_version != undef and versioncmp("${::aio_agent_version}", "${_package_version}") >= 0 {
```

Comment:
That makes sense to me. I think the only `is_pe` checks we should be doing are for the service check (since PE already manages the services) and for the PE only platforms (AIX, etc)